### PR TITLE
DS-172 - feat/DS-172/snapshot-calculator

### PR DESCRIPTION
**Pull request by knzeng-e**
This PR adds the ability to create a JSON file containing a list of DIDs with given staking on a certain day.

### CODE REVIEW

**Comment by knzeng-e** [comment]
in commit [b35589b](https://github.com/energywebfoundation/staking-pool/pull/48/commits/b35589b5ee35fbe4d42bf036ea99dadc048acbb4), I was thinking that we would probably need to query all events from the block 0 (or at least the block from which the staking contract where deployed).

The reason for this is that since we are reading `staking` event logs, in a scenario where we are taking a snapshot from a later block, one may have staked before (so the corresponding event wont be in this scope) and hence would not have his/her stake referred in this snapshot batch, even if he/she didn't unstaked.

Does that make sens ?

File: scripts/snapshot-calculator.ts [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+        if (failingCalculations.includes(currentLog)) {
+          failingCalculations = failingCalculations.filter((log) => {
+            return log.address !== currentLog.address;
+          });
+        }
+        return stakers;
+      } catch (error) {
+        console.log(`An error occurred during calculation of snapshot ${currentAddress} ...`);
+        //We keep track of all transactions which fail
+        if (!failingCalculations.includes(currentLog)) {
```
Comment (JGiter) [inline scripts/snapshot-calculator.ts:115]: It seems that this way of keeping track of failing calculations isn't very useful as they are not handled but just dropped after script is finished. Am I missing something?

File: scripts/snapshot-calculator.ts [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+      //Removing the padding zeros
+      const currentAddress = `0x${log.topics[1].substring(26)}`;
+      const currentLog = {
+        address: currentAddress,
+        blockNumber: log.blockNumber,
+        transactionHash: log.transactionHash,
+      };
+
+      //Rretrieving addresses and avoiding dupplication
+      if (!stakingLogs.includes(currentLog)) {
```
[the remark this answers is not in the document]
Comment (knzeng-e) [reply scripts/snapshot-calculator.ts:42]: Oh I see  thanks ! it makes sense, and it also explains why I still had some logs duplicated (hence the use of Set type here >> https://github.com/energywebfoundation/staking-pool/blob/b35589b5ee35fbe4d42bf036ea99dadc048acbb4/scripts/snapshot-calculator.ts#L156.

Thank you @JGiter

File: scripts/snapshot-calculator.ts (hunk 1 above)
Comment (knzeng-e) [reply scripts/snapshot-calculator.ts:115]: At this step, those failing calculation are indeed just added into this global array. But they are handled in a second time inside the `takeSnapshot` function; after calling the `calculateSnapshot` function (the one here where the failing calculations are listed), we will keep calling the `calculateSnapshot` on this `falingCalculations` array while is they are not all handled.

Please have a look here >> https://github.com/energywebfoundation/staking-pool/blob/b35589b5ee35fbe4d42bf036ea99dadc048acbb4/scripts/snapshot-calculator.ts#L149

Comment (JGiter) [reply scripts/snapshot-calculator.ts:115]: You right. I missed this. Thanks

File: scripts/snapshot-calculator.ts [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    address: stakingPoolAddress,
+    topics: ["0x270d6dd254edd1d985c81cf7861b8f28fb06b6d719df04d90464034d43412440"], // Ref of topic corresponding to `StakeAdded` event
+    fromBlock: fromBlockNumber, //should it always be 0 ?
+    toBlock: toBlockNumber,
+  });
+
+  const allStakers = await parseEvents(logs);
+
+  return allStakers;
+};
```
Comment (jrhender) [inline scripts/snapshot-calculator.ts:78]: I think we might also need to get the "stake removed" events to sum the stakers stake at a given block. Is this being done somewhere and I just missed it?

File: scripts/deploy-patron-kyc.js [hunk 4]
```diff
@@ -35,14 +34,18 @@ const deployContract = async (contractName) => {
...
-  const end = start + 24 * 3600;
+  const start = 1647630000; //18/03/2022 - 20:00:00
+  const end = 1671390000; //18/12/2022 - 20:00:0
+  const ratio = ethers.utils.parseUnits("0.00001125", 18);
+
+  const hardCap = ethers.utils.parseUnits("7500000", "ether");
+  const contributionLimit = ethers.utils.parseUnits("3000", "ether");
+
+  // const patronRoles = [ethers.utils.namehash("email.roles.verification.apps.energyweb.iam.ewc")]; //on volta
+  const patronRoles = [ethers.utils.namehash("email.roles.verification.apps.energyweb.auth.ewc")]; // EWC
```
[the remark this answers is not in the document]
Comment (knzeng-e) [reply scripts/deploy-patron-kyc.js:47]: Oh this was a previous commit from the branch I checked out. Sorry I should have stashed it before

File: scripts/snapshot-calculator.ts (hunk 3 above)
Comment (knzeng-e) [reply scripts/snapshot-calculator.ts:78]: Good point, it is actually missing, and this is related to the misunderstanding I had, and for which I raised the question here >> https://energyweb.atlassian.net/browse/DS-172?focusedCommentId=20834. Would you please check it, your input will help for sure. Thank you @jrhender

**Review by jrhender — COMMENTED** [review]
Thanks @knzeng-e ! Looks like some great progress. Could you add some tests for the follow scenarios please:
1. should not create a snapshot if a staker stakes lower than the snapshot minimum
2. should not create a snapshot if a staker stakes higher than the snapshot minimum but withdraws before the snapshot block

File: scripts/snapshot-calculator.ts [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+import { Log } from "@ethersproject/providers";
+import { Contract, Wallet, providers } from "ethers";
+import { Snapshot } from "./types/snapshot.types";
+import { getRpc, EW_CHAIN_ID, formatDID, _rpcReadContractSlot } from "./utils/snapshot.utils";
+
+config();
+let stakingContract: Contract;
+let failingCalculations: string[] = [];
+
+const STAKES_STORAGE_SLOT = 12;
```
[the remark this answers is not in the document]
Comment (knzeng-e) [reply scripts/snapshot-calculator.ts:13]: I was reading this from remix IDE ([compilation details](https://remix-ide.readthedocs.io/en/stable/compile.html#compilation-details-and-publishing)), but I will add a way to retrieve this from the generated artifacts (produced during compilation)

File: test/StakingPoolPatronKYC.test.ts [hunk 6]
```diff
@@ -436,13 +443,36 @@ describe("Staking Pool Patron KYC", function () {
...
+
+    it("should not create a snapshot if a staker stakes lower than the snapshot minimum", async () => {
+      const { asPatron3, provider, stakingPool } = await loadFixture(defaultFixture);
+
+      tx = await asPatron3.stake({ value: oneEWT.mul(2) });
+      await tx.wait();
+      lastBlock = await provider.getBlock("latest");
+      const snapshot = await takeSnapShot(stakingPool.address, chainID, lastBlock.number, 5, provider);
+      expect(snapshot).to.be.null;
+    });
```
Comment (jrhender) [inline test/StakingPoolPatronKYC.test.ts:464]: To me, this test would be better if it had:
1. both the current staker who lowers their balance and also a staker who does have a high enough balance. I want to avoid this behaviour (which I think would pass the current test cases): If one or more stakers don't have the minimum balance, then no file is output.
2. the staker who stakers lower (prior to the snapshot block) then stakes higher after the snapshot block doesn't get the snapshot

Probably point 2 deserves a separate test, IMO

File: test/StakingPoolPatronKYC.test.ts [hunk 7]
```diff
@@ -436,13 +443,36 @@ describe("Staking Pool Patron KYC", function () {
...
+      const { asPatron3, provider, stakingPool } = await loadFixture(defaultFixture);
+      const minBalance = 15;
+
+      await asPatron3.stake({ value: oneEWT.mul(50) });
+      await asPatron3.unstake(oneEWT.mul(49));
+      lastBlock = await provider.getBlock("latest");
+
+      const _snapshot = await takeSnapShot(stakingPool.address, chainID, lastBlock.number, minBalance, provider);
+      expect(_snapshot).to.be.null;
+    });
```
Comment (jrhender) [inline test/StakingPoolPatronKYC.test.ts:475]: Similar to my other comment, I think this test would give me more confidence, if it tested both the "negative" and the "positive" scenario at once. In other words, I would also add a staker that does met the requirement to this test.

Also like the other comment, I think it would be good to add a separate test where the staker re-stakes after the snapshot block

**Review by jrhender — COMMENTED** [review]
Some further improvement suggestions for the tests

**Comment by jrhender** [comment]
@knzeng-e looking at the file issuer documentation, I think it is expecting `snapshotBlock` in the `issuerFields`. Does that make sense to you? If so, could you add please 🙇 ?

File: test/StakingPoolPatronKYC.test.ts [hunk 8]
```diff
@@ -453,15 +453,83 @@ describe("Staking Pool Patron KYC", function () {
       expect(snapshotRetro).to.equal(expectedSnapshot1);
     });
 
+    it("should not include a DID which stakes lower than the minimum", async () => {
```
Comment (jrhender) [inline test/StakingPoolPatronKYC.test.ts:456]: Just adjusting the title a bit to differentiate from the "should not create a snapshot if a staker stakes lower than the snapshot minimum" test
```suggestion
    it("should not include a staker who stakes the minimum after the snapshot", async () => {
```

File: scripts/snapshot-calculator.ts [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+
+const getStakers = async (
+  blockNumber: number,
+  minStakedAmount: number,
+  stakingPoolAddress: string,
+  provider: JsonRpcProvider,
+) => {
+  stakingContract = (await ethers.getContractFactory("StakingPoolPatronKYC", wallet.connect(provider))).attach(
+    stakingPoolAddress,
+  );
```
Comment (jrhender) [inline scripts/snapshot-calculator.ts:42]: Is this necessary? It seems it is unused
```suggestion
```

File: scripts/deploy-patron-kyc.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (jrhender) [reply scripts/deploy-patron-kyc.js:47]: @knzeng-e ok, do we still want to include the changes in this PR? Would it be possible to include them in a separate PR please?

File: hardhat.config.ts [hunk 10]
```diff
@@ -15,8 +15,22 @@ const deployer_privateKey = process.env.DEPLOYER_PRIV_KEY || defaultAccounts[0].
...
 module.exports = {
-  solidity: "0.8.6",
   defaultNetwork: "hardhat",
+  solidity: {
+    version: "0.8.6",
+    settings: {
+      optimizer: {
+        enabled: true,
+        runs: 200,
+      },
```
Comment (jrhender) [inline hardhat.config.ts:25]: Why this change? I guess this would alter the bytecode output. If so, IMO, we should not do this as a part of this PR

**Review by jrhender — COMMENTED** [review]
Still reviewing but some posting further comments. Mostly style/documentation comments though, I think we're nearly there!

File: scripts/snapshot-calculator.ts [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+            });
+          }
+        }
+        return snapshots;
+      } catch (error) {
+        console.log(`An error occurred during calculation of snapshot ${currentAddress} ... `);
+        // We keep track of all transactions which fail
+        if (!failingCalculations.includes(currentAddress)) {
+          failingCalculations.push(currentAddress);
+        }
```
Comment (jrhender) [inline scripts/snapshot-calculator.ts:111]: For this behaviour of "add if it isn't in collection", it would be better to use a `Set` IMO as this is the job it is designed for and it simplifies the code a bit https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set
```suggestion
        failingCalculations.add(currentAddress);
```

File: scripts/snapshot-calculator.ts [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+          if (failingCalculations.includes(currentAddress)) {
+            failingCalculations = failingCalculations.filter((addr) => {
+              return addr !== currentAddress;
+            });
+          }
+        }
+        return snapshots;
+      } catch (error) {
+        console.log(`An error occurred during calculation of snapshot ${currentAddress} ... `);
+        // We keep track of all transactions which fail
```
Comment (jrhender) [inline scripts/snapshot-calculator.ts:108]: This comment is a bit misleading IMO as there aren't transactions as not blockchain state is changing. Would you agree with this @knzeng-e ?
```suggestion
        // We keep track of all snapshot calculations which fail
```

File: scripts/snapshot-calculator.ts [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+const parseEvents = async (stakeLogs: Log[], blockNumber: number) => {
+  const stakers: string[] = [];
+  await Promise.all(
+    stakeLogs.map(async (currentLog) => {
+      if (currentLog.blockNumber <= blockNumber) {
+        //Removing the padding zeros
+        const currentAddress = utils.hexStripZeros(currentLog.topics[1]);
+        if (!stakers.includes(currentAddress)) {
+          stakers.push(currentAddress);
+        }
```
Comment (jrhender) [inline scripts/snapshot-calculator.ts:27]: Similar to my other comment, better to use a `Set` here
```suggestion
        stakers.add(currentAddress)
```

**Review by jrhender — COMMENTED** [review]
Suggestions to use `Set`

File: scripts/snapshot-calculator.ts (hunk 12 above)
Comment (knzeng-e) [reply scripts/snapshot-calculator.ts:108]: oh yes, I agree @jrhender !

File: hardhat.config.ts (hunk 10 above)
Comment (knzeng-e) [reply hardhat.config.ts:25]: right, I forgot to revert this change in fact. II had an issue with the generation of my artifacts on compilation, I wanted to force-check major change. Thanks @jrhender

File: scripts/snapshot-calculator.ts [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+      }
+    }),
+  );
+  return stakers;
+};
+
+const getStakers = async (
+  blockNumber: number,
+  minStakedAmount: number,
```
[the remark this answers is not in the document]
Comment (knzeng-e) [reply scripts/snapshot-calculator.ts:36]: right !

File: scripts/snapshot-calculator.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (knzeng-e) [reply scripts/snapshot-calculator.ts]: update: implemented [here](https://github.com/energywebfoundation/staking-pool/blob/e22ce5a556fc1f6f96f3159844857a6e82a130f7/scripts/utils/snapshot.utils.ts#L41)

File: scripts/snapshot-calculator.ts [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+      minimumBalance,
+      contractAddress,
+    );
+    snapshotContent = snapshotContent.concat(failingSnapshot);
+  }
+
+  const finalSnapshot = [...new Set(snapshotContent)];
+
+  const snapshot = {
+    credentialNamespace: "snapshot1.roles.consortiapool.apps.energyweb.iam.ewc",
```
[the remark this answers is not in the document]
Comment (knzeng-e) [reply scripts/snapshot-calculator.ts:144]: ok @jrhender , I'm on it

File: scripts/deploy-patron-kyc.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (knzeng-e) [reply scripts/deploy-patron-kyc.js:47]: Apologizes, I didn't see this last comment. I remove this yes

**Review by jrhender — APPROVED** [review]
Looks great, thanks for all your efforts @knzeng-e !
